File: ruby/quicksort.rb
module Sorting
  def self.quicksort(array)
    return array if array.length <= 1

    index = [((array.length - 1) / 2).floor, 0].max
    pivot = array[index]

    left = []
    right = []
    pivot_a = []
    array.each do |i|
      if i < pivot
        left << i
      elsif i > pivot
        right << i
      else 
        pivot_a << i
      end
    end
    quicksort(left) + pivot_a + quicksort(right)
  end

  def self.in_place_quicksort(array, start_i = 0, end_i = nil)
    end_i ||= array.length - 1

    return array if start_i >= end_i

    pivot_i = [((end_i - start_i)/2).floor, start_i].max
    
    left = start_i
    right = end_i 

    while left <= right 
      while array[left] < array[pivot_i]
        left += 1
      end

      while array[right] > array[pivot_i]
        right -= 1
      end

      if left <= right
        tmp = array[left]
        array[left] = array[right]
        array[right] = tmp

        left += 1
        right -= 1
      end
    end

    in_place_quicksort(array, start_i, right)
    in_place_quicksort(array, left, end_i)
  end
end

File: erlang/quicksort.erl
% Compile with: erl quicksort.erl
% Run with: erl -noshell -s quicksort test -s init stop

-module(quicksort).
-include_lib("eunit/include/eunit.hrl").
-export([quicksort/1]).

quicksort([]) -> [];
quicksort([I]) -> [I];
quicksort(Array) -> 
	Pivot_i = trunc(length(Array) / 2),
	Pivot = lists:nth(Pivot_i, Array),
	quicksort([Left || Left <- Array, Left < Pivot]) ++ [P || P <- Array, P == Pivot] ++ quicksort([Right || Right <- Array, Right > Pivot]).

%%
%% Tests
quicksort_test() -> 
	Arr = [10, 2, 1, 7, 9, 8, 0, 4, 6, 5, 3],
	[0,1,2,3,4,5,6,7,8,9,10] = quicksort(Arr).

quicksort_empty_test() -> [] = quicksort([]).
quicksort_single_test() -> [3] = quicksort([3]).
quicksort_ordered_test() -> 
	Arr = [0,1,2,3,4,5,6,7,8,9,10],
	[0,1,2,3,4,5,6,7,8,9,10] = quicksort(Arr).
